In [251]:
import numpy as np
from collections import defaultdict

**1.** Make a solution to generate a stream of random numbers `(amount = 1 000 000)` with values in interval `[0..1000]` with delay time = 0.2 of second (to avoid overhead).

In [252]:
np.random.seed(42)

In [253]:
MAX_AMOUNT = 1_000_000

In [254]:
stream = np.array([np.random.randint(0, 1000 + 1) for _ in range(MAX_AMOUNT)])

**2.** Calculate `0-th moment` -- a count of the number of distinct elements in the stream.

In [255]:
freqs = defaultdict(int)
for x in stream:
    freqs[x] += 1

In [256]:
m_0 = len(freqs.keys())
print(f'0-th moment: {m_0}')

0-th moment: 1001


**3.** Calculate `1-th moment` -- a count of the number of all elements in the stream.

In [257]:
m_1 = sum(freqs.values())
print(f'1-th moment: {m_1}')

1-th moment: 1000000


**4.** Calculate `2-nd moment` based on the `"Alon-Matias-Szegedy" Algorithm`

    a) with 100 variables
    b) with 500 variables


In [258]:
m_2 = sum([x * x for x in freqs.values()])
print(f'True 2-nd moment: {m_2}')

True 2-nd moment: 999945184


a) with 100 variables

In [259]:
variables_1 = defaultdict(int)

In [260]:
# 100 random positions in a stream
while len(variables_1) != 100:
    variables_1[np.random.randint(0, MAX_AMOUNT)] = 0

In [261]:
for i in variables_1.keys():
    for j in range(i, len(stream)):
        if stream[i] == stream[j]:
            variables_1[i] += 1

Let's estimate 2nd moment by ${n * (2 * X_{.value} - 1)}$ and take the average

In [262]:
values = np.array(list(variables_1.values()))
estimate_m_2 = np.mean(((2 * values) - 1)) * MAX_AMOUNT
print(f'Estimate 2nd moment by 100 variables: {estimate_m_2}')

Estimate 2nd moment by 100 variables: 971980000.0


In [266]:
percentage_difference = (estimate_m_2 / m_2 - 1) * 100
percentage_difference

-2.796671702356035

The `estimate 2nd moment` calculated with 100 variables is less than the `true 2 nd moment` by $\approx 2.8$ %

b) with 500 variables

In [267]:
variables_2 = defaultdict(int)

In [268]:
# 500 random positions in a stream
while len(variables_2) != 500:
    variables_2[np.random.randint(0, MAX_AMOUNT)] = 0

In [269]:
for i in variables_2.keys():
    for j in range(i, len(stream)):
        if stream[i] == stream[j]:
            variables_2[i] += 1

Let's estimate 2nd moment by ${n * (2 * X_{.value} - 1)}$ and take the average

In [270]:
values = np.array(list(variables_2.values()))
estimate_m_2 = np.mean(((2 * values) - 1)) * MAX_AMOUNT
print(f'Estimate 2nd moment by 500 variables: {estimate_m_2}')

Estimate 2nd moment by 500 variables: 1013172000.0


In [271]:
percentage_difference = (estimate_m_2 / m_2 - 1) * 100
percentage_difference

1.322754108089197

The `estimate 2nd moment` calculated with 500 variables is more than the `true 2 nd moment` by $\approx 1.3$ %